lib/saluki-components/src/transforms/dogstatsd_mapper/mod.rs

```diff
@@ -0,0 +1,254 @@
+use std::collections::HashMap;
+use std::num::NonZeroUsize;
+use std::str::FromStr;
+use std::sync::LazyLock;
+
+use async_trait::async_trait;
+use bytesize::ByteSize;
+use memory_accounting::{MemoryBounds, MemoryBoundsBuilder};
+use regex::Regex;
+use saluki_config::GenericConfiguration;
+use saluki_context::{Context, ContextResolver, ContextResolverBuilder};
+use saluki_core::{
+    components::transforms::{SynchronousTransform, SynchronousTransformBuilder},
+    topology::interconnect::FixedSizeEventBuffer,
+};
+use saluki_error::{generic_error, GenericError};
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, DisplayFromStr, PickFirst};
+
+pub const MATCH_TYPE_WILDCARD: &str = "wildcard";
+pub const MATCH_TYPE_REGEX: &str = "regex";
+
+static ALLOWED_WILDCARD_MATCH_PATTERN: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"^[a-zA-Z0-9\-_*.]+$").expect("Invalid regex in ALLOWED_WILDCARD_MATCH_PATTERN"));
+
+/// DogstatsD mapper transform.
+#[serde_as]
+#[derive(Deserialize)]
+pub struct DogstatsDMapperConfiguration {
+    #[serde(skip)]
+    context_string_interner_bytes: ByteSize,
+
+    #[serde_as(as = "PickFirst<(DisplayFromStr, _)>")]
+    #[serde(default)]
+    dogstatsd_mapper_profiles: MapperProfileConfigs,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+struct MapperProfileConfigs(pub Vec<MappingProfileConfig>);
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+struct MappingProfileConfig {
+    pub name: String,
+    pub prefix: String,
+    pub mappings: Vec<MetricMappingConfig>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+struct MetricMappingConfig {
+    #[serde(rename = "match")]
+    metric_match: String,
+    #[serde(rename = "match_type")]
+    match_type: String,
+    name: String,
+    tags: HashMap<String, String>,
+}
+
+struct MetricMapper {
+    profiles: Vec<MappingProfile>,
+}
+
+impl MetricMapper {
+    fn map(&self, context_resolver: &mut ContextResolver, metric_name: &str) -> Option<Context> {
+        for profile in &self.profiles {
+            if !metric_name.starts_with(&profile.prefix) && profile.prefix != "*" {
+                continue;
+            }
+
+            for mapping in &profile.mappings {
+                if let Some(captures) = mapping.regex.captures(metric_name) {
+                    let mut name = String::new();
+                    captures.expand(&mapping.name, &mut name);
+
+                    let mut tags = Vec::with_capacity(mapping.tags.len());
+                    for (tag_key, tag_value_expr) in &mapping.tags {
+                        let mut expanded_value = String::new();
+                        captures.expand(tag_value_expr, &mut expanded_value);
+                        tags.push(format!("{}:{}", tag_key, expanded_value));
+                    }
+                    return context_resolver.resolve(&name, tags, None);
```

Comment:
So two things jump out at me here:

- we're ignoring the existing tags on the metric and just keeping whatever new tags we generate from the metric name
- we're going to lose origin tags since we're not doing anything with keeping the origin info

For the first one, and perhaps the second one: does the Agent actually throw away the existing tags on the metric in the process of mapping it?

If so, then this code would be fine as-is, I think? If not, then we'd obviously need to clone the existing tags, but we'd also need to copy over the origin information... and we don't really have a way to do that currently, so I'm going to think on that a little bit.

Reply:
Oops thanks for catching this. The new tags should be added to the existing tags. Will update this